x/programs/cmd/simulator/src/simulator.rs

```diff
@@ -0,0 +1,367 @@
+            .ok_or(SimulatorError::Client(ClientError::Eof))?
+            .and_then(|request| {
+                if let Some(err) = request.error {
+                    Err(SimulatorError::Program(err))
+                } else {
+                    Ok(request)
+                }
+            })
+    }
+
+    pub fn execute(
+        &mut self,
+        method: String,
+        params: Vec<Param>,
+        max_units: u64,
+    ) -> SimulatorResponseItem {
+        self.writer.write_all(Self::RUN_COMMAND)?;
+
+        let input = serde_json::to_vec(&SimulatorRequest::new_execute(method, params, max_units))?;
+
+        self.writer.write_all(&input)?;
+        self.writer.write_all(b"'\n")?;
+        self.writer.flush()?;
+
+        self.responses
+            .next()
+            .ok_or(SimulatorError::Client(ClientError::Eof))?
+            .and_then(|request| {
+                if let Some(err) = request.error {
+                    Err(SimulatorError::Program(err))
+                } else {
+                    Ok(request)
+                }
+            })
+    }
+}
+
+/// A [`SimulatorRequest`] is a call to the simulator
+#[derive(Debug, Serialize, PartialEq, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct SimulatorRequest {
+    /// The API endpoint to call.
+    pub endpoint: Endpoint,
+    /// The method to call on the endpoint.
+    pub method: String,
+    /// The maximum number of units the Request can consume.
+    pub max_units: u64,
+    /// The parameters to pass to the method.
+    pub params: Vec<Param>,
+}
+
+impl SimulatorRequest {
+    pub fn new_read(method: String, params: Vec<Param>) -> Self {
+        Self {
+            endpoint: Endpoint::ReadOnly,
+            method: method.to_string(),
+            max_units: 0,
+            params,
+        }
+    }
+
+    pub fn new_execute(method: String, params: Vec<Param>, max_units: u64) -> Self {
+        Self {
+            endpoint: Endpoint::Execute,
+            method,
+            max_units,
+            params,
+        }
+    }
+
+    pub fn new_create_program(path: String) -> Self {
+        Self {
+            endpoint: Endpoint::CreateProgram,
+            // TODO: this does not need to be passed in
+            method: "create_program".to_string(),
+            max_units: 0,
+            params: vec![Param::Path(path)],
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum SimulatorError {
+    #[error("Client error: {0}")]
+    Client(#[from] ClientError),
+    #[error("Serialization / Deserialization error: {0}")]
+    Serde(#[from] serde_json::Error),
+    #[error("Borsh deserialization error: {0}")]
+    BorshDeserialization(#[from] borsh::io::Error),
+    #[error("Program error: {0}")]
+    Program(String),
+}
+pub(crate) type SimulatorResponseItem = Result<SimulatorResponse, SimulatorError>;
+
+#[derive(Debug, Deserialize)]
+pub struct SimulatorResult {
+    /// The ID created from the program execution.
+    pub action_id: Option<String>,
```

Comment:
@iFrostizz, did you put this here or @dboehm-avalabs? Why is it here? It's not used if I'm not mistaken.

Reply:
git blamed @dboehm-avalabs ! https://github.com/ava-labs/hypersdk/pull/1056
I can't find of any really good use-case besides some funny PoW stuff

Reply:
Oh wow that's totally unrelated ^ this is for the context !
No it was a legacy code, but we don't seem to need that. And it could simplified to just have the `result`, so that the `response` is now just a method of the `SimulatorResponse`.